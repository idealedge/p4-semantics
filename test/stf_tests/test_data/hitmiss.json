{
  "program" : "hitmiss.p4",
  "__meta__" : {
    "version" : [2, 7],
    "compiler" : "https://github.com/p4lang/p4c"
  },
  "header_types" : [
    {
      "name" : "scalars_0",
      "id" : 0,
      "fields" : [
        ["tmp_0", 1, false],
        ["_padding_0", 7, false]
      ]
    },
    {
      "name" : "data_t",
      "id" : 1,
      "fields" : [
        ["f1", 32, false],
        ["f2", 32, false],
        ["f3", 32, false],
        ["f4", 32, false],
        ["b1", 8, false],
        ["b2", 8, false],
        ["h1", 16, false]
      ]
    },
    {
      "name" : "standard_metadata",
      "id" : 2,
      "fields" : [
        ["ingress_port", 9, false],
        ["egress_spec", 9, false],
        ["egress_port", 9, false],
        ["clone_spec", 32, false],
        ["instance_type", 32, false],
        ["drop", 1, false],
        ["recirculate_port", 16, false],
        ["packet_length", 32, false],
        ["enq_timestamp", 32, false],
        ["enq_qdepth", 19, false],
        ["deq_timedelta", 32, false],
        ["deq_qdepth", 19, false],
        ["ingress_global_timestamp", 48, false],
        ["lf_field_list", 32, false],
        ["mcast_grp", 16, false],
        ["resubmit_flag", 1, false],
        ["egress_rid", 16, false],
        ["checksum_error", 1, false],
        ["_padding", 4, false]
      ]
    }
  ],
  "headers" : [
    {
      "name" : "scalars",
      "id" : 0,
      "header_type" : "scalars_0",
      "metadata" : true,
      "pi_omit" : true
    },
    {
      "name" : "standard_metadata",
      "id" : 1,
      "header_type" : "standard_metadata",
      "metadata" : true,
      "pi_omit" : true
    },
    {
      "name" : "data",
      "id" : 2,
      "header_type" : "data_t",
      "metadata" : false,
      "pi_omit" : true
    }
  ],
  "header_stacks" : [],
  "header_union_types" : [],
  "header_unions" : [],
  "header_union_stacks" : [],
  "field_lists" : [],
  "errors" : [
    ["NoError", 1],
    ["PacketTooShort", 2],
    ["NoMatch", 3],
    ["StackOutOfBounds", 4],
    ["HeaderTooShort", 5],
    ["ParserTimeout", 6]
  ],
  "enums" : [],
  "parsers" : [
    {
      "name" : "parser",
      "id" : 0,
      "init_state" : "start",
      "parse_states" : [
        {
          "name" : "start",
          "id" : 0,
          "parser_ops" : [
            {
              "parameters" : [
                {
                  "type" : "regular",
                  "value" : "data"
                }
              ],
              "op" : "extract"
            }
          ],
          "transitions" : [
            {
              "value" : "default",
              "mask" : null,
              "next_state" : null
            }
          ],
          "transition_key" : []
        }
      ]
    }
  ],
  "deparsers" : [
    {
      "name" : "deparser",
      "id" : 0,
      "order" : ["data"]
    }
  ],
  "meter_arrays" : [],
  "counter_arrays" : [],
  "register_arrays" : [],
  "calculations" : [],
  "learn_lists" : [],
  "actions" : [
    {
      "name" : "NoAction",
      "id" : 0,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "NoAction",
      "id" : 1,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "NoAction",
      "id" : 2,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "setb1",
      "id" : 3,
      "runtime_data" : [
        {
          "name" : "val",
          "bitwidth" : 8
        },
        {
          "name" : "port",
          "bitwidth" : 9
        }
      ],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["data", "b1"]
            },
            {
              "type" : "runtime_data",
              "value" : 0
            }
          ],
          "source_info" : {
            "filename" : "hitmiss.p4",
            "line" : 37,
            "column" : 13,
            "source_fragment" : "val, port) { ..."
          }
        },
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["standard_metadata", "egress_spec"]
            },
            {
              "type" : "runtime_data",
              "value" : 1
            }
          ],
          "source_info" : {
            "filename" : "hitmiss.p4",
            "line" : 37,
            "column" : 18,
            "source_fragment" : "port) { ..."
          }
        }
      ]
    },
    {
      "name" : "setb1",
      "id" : 4,
      "runtime_data" : [
        {
          "name" : "val",
          "bitwidth" : 8
        },
        {
          "name" : "port",
          "bitwidth" : 9
        }
      ],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["data", "b1"]
            },
            {
              "type" : "runtime_data",
              "value" : 0
            }
          ],
          "source_info" : {
            "filename" : "hitmiss.p4",
            "line" : 37,
            "column" : 13,
            "source_fragment" : "val, port) { ..."
          }
        },
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["standard_metadata", "egress_spec"]
            },
            {
              "type" : "runtime_data",
              "value" : 1
            }
          ],
          "source_info" : {
            "filename" : "hitmiss.p4",
            "line" : 37,
            "column" : 18,
            "source_fragment" : "port) { ..."
          }
        }
      ]
    },
    {
      "name" : "setb1",
      "id" : 5,
      "runtime_data" : [
        {
          "name" : "val",
          "bitwidth" : 8
        },
        {
          "name" : "port",
          "bitwidth" : 9
        }
      ],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["data", "b1"]
            },
            {
              "type" : "runtime_data",
              "value" : 0
            }
          ],
          "source_info" : {
            "filename" : "hitmiss.p4",
            "line" : 37,
            "column" : 13,
            "source_fragment" : "val, port) { ..."
          }
        },
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["standard_metadata", "egress_spec"]
            },
            {
              "type" : "runtime_data",
              "value" : 1
            }
          ],
          "source_info" : {
            "filename" : "hitmiss.p4",
            "line" : 37,
            "column" : 18,
            "source_fragment" : "port) { ..."
          }
        }
      ]
    },
    {
      "name" : "noop",
      "id" : 6,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "noop",
      "id" : 7,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "noop",
      "id" : 8,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "act",
      "id" : 9,
      "runtime_data" : [],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["scalars", "tmp_0"]
            },
            {
              "type" : "expression",
              "value" : {
                "type" : "expression",
                "value" : {
                  "op" : "b2d",
                  "left" : null,
                  "right" : {
                    "type" : "bool",
                    "value" : true
                  }
                }
              }
            }
          ]
        }
      ]
    },
    {
      "name" : "act_0",
      "id" : 10,
      "runtime_data" : [],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["scalars", "tmp_0"]
            },
            {
              "type" : "expression",
              "value" : {
                "type" : "expression",
                "value" : {
                  "op" : "b2d",
                  "left" : null,
                  "right" : {
                    "type" : "bool",
                    "value" : false
                  }
                }
              }
            }
          ]
        }
      ]
    }
  ],
  "pipelines" : [
    {
      "name" : "ingress",
      "id" : 0,
      "init_table" : "test1",
      "tables" : [
        {
          "name" : "test1",
          "id" : 0,
          "source_info" : {
            "filename" : "hitmiss.p4",
            "line" : 42,
            "column" : 0,
            "source_fragment" : "table test1 { ..."
          },
          "key" : [
            {
              "match_type" : "ternary",
              "target" : ["data", "f1"],
              "mask" : null
            }
          ],
          "match_type" : "ternary",
          "type" : "simple",
          "max_size" : 1024,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [3, 6, 0],
          "actions" : ["setb1", "noop", "NoAction"],
          "base_default_next" : null,
          "next_tables" : {
            "__HIT__" : "tbl_act",
            "__MISS__" : "tbl_act_0"
          },
          "default_entry" : {
            "action_id" : 0,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        },
        {
          "name" : "tbl_act",
          "id" : 1,
          "key" : [],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 1024,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [9],
          "actions" : ["act"],
          "base_default_next" : "node_5",
          "next_tables" : {
            "act" : "node_5"
          },
          "default_entry" : {
            "action_id" : 9,
            "action_const" : true,
            "action_data" : [],
            "action_entry_const" : true
          }
        },
        {
          "name" : "tbl_act_0",
          "id" : 2,
          "key" : [],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 1024,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [10],
          "actions" : ["act_0"],
          "base_default_next" : "node_5",
          "next_tables" : {
            "act_0" : "node_5"
          },
          "default_entry" : {
            "action_id" : 10,
            "action_const" : true,
            "action_data" : [],
            "action_entry_const" : true
          }
        },
        {
          "name" : "test2",
          "id" : 3,
          "source_info" : {
            "filename" : "hitmiss.p4",
            "line" : 52,
            "column" : 0,
            "source_fragment" : "table test2 { ..."
          },
          "key" : [
            {
              "match_type" : "ternary",
              "target" : ["data", "f2"],
              "mask" : null
            }
          ],
          "match_type" : "ternary",
          "type" : "simple",
          "max_size" : 1024,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [4, 7, 1],
          "actions" : ["setb1", "noop", "NoAction"],
          "base_default_next" : null,
          "next_tables" : {
            "setb1" : null,
            "noop" : null,
            "NoAction" : null
          },
          "default_entry" : {
            "action_id" : 1,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        },
        {
          "name" : "test3",
          "id" : 4,
          "source_info" : {
            "filename" : "hitmiss.p4",
            "line" : 62,
            "column" : 0,
            "source_fragment" : "table test3 { ..."
          },
          "key" : [
            {
              "match_type" : "ternary",
              "target" : ["data", "f3"],
              "mask" : null
            }
          ],
          "match_type" : "ternary",
          "type" : "simple",
          "max_size" : 1024,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [5, 8, 2],
          "actions" : ["setb1", "noop", "NoAction"],
          "base_default_next" : null,
          "next_tables" : {
            "setb1" : null,
            "noop" : null,
            "NoAction" : null
          },
          "default_entry" : {
            "action_id" : 2,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        }
      ],
      "action_profiles" : [],
      "conditionals" : [
        {
          "name" : "node_5",
          "id" : 0,
          "expression" : {
            "type" : "expression",
            "value" : {
              "op" : "d2b",
              "left" : null,
              "right" : {
                "type" : "field",
                "value" : ["scalars", "tmp_0"]
              }
            }
          },
          "true_next" : "test2",
          "false_next" : "test3"
        }
      ]
    },
    {
      "name" : "egress",
      "id" : 1,
      "init_table" : null,
      "tables" : [],
      "action_profiles" : [],
      "conditionals" : []
    }
  ],
  "checksums" : [],
  "force_arith" : [],
  "extern_instances" : [],
  "field_aliases" : [
    [
      "queueing_metadata.enq_timestamp",
      ["standard_metadata", "enq_timestamp"]
    ],
    [
      "queueing_metadata.enq_qdepth",
      ["standard_metadata", "enq_qdepth"]
    ],
    [
      "queueing_metadata.deq_timedelta",
      ["standard_metadata", "deq_timedelta"]
    ],
    [
      "queueing_metadata.deq_qdepth",
      ["standard_metadata", "deq_qdepth"]
    ],
    [
      "intrinsic_metadata.ingress_global_timestamp",
      ["standard_metadata", "ingress_global_timestamp"]
    ],
    [
      "intrinsic_metadata.lf_field_list",
      ["standard_metadata", "lf_field_list"]
    ],
    [
      "intrinsic_metadata.mcast_grp",
      ["standard_metadata", "mcast_grp"]
    ],
    [
      "intrinsic_metadata.resubmit_flag",
      ["standard_metadata", "resubmit_flag"]
    ],
    [
      "intrinsic_metadata.egress_rid",
      ["standard_metadata", "egress_rid"]
    ]
  ]
}