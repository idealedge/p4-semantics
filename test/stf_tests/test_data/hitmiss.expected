00000000000000000000000100000001************************************************************************************************0111111101100110 2
00000000000000000000001000000010************************************************************************************************0000011101100110 3
0000000000000000000000010000000100100010001000100010001000100010000000000000000000000011000000110000000000000000000001000000010000000001011001100111011110001000 4
0000000000000000000100010001000100100010001000100010001000100010000000000000000000000011000000110000000000000000000001000000010000000010011001100111011110001000 5